### PR TITLE
Inroduce `service-provider` module

### PR DESCRIPTION
**Pull request by alllex**
No code changes, only moves some classes into a new module.

Adds a new module that is home to the classes that allows other modules to register internal Gradle services.
This includes classes like `ServiceRegistration` and `PluginServiceRegistry`.

The `Scope` interfaces should later be moved here as well, because they participate in the service declaration.

### CODE REVIEW

**Comment by alllex** [comment]
@bot-gradle test this

File: platforms/core-configuration/declarative-dsl-tooling-builders/build.gradle.kts [hunk 1]
```diff
@@ -5,9 +5,9 @@ plugins {
 description = "Declarative DSL Tooling Builders for IDEs"
 
 dependencies {
-    api(project(":base-services"))
     api(project(":core"))
     api(project(":core-api"))
+    api(project(":service-provider"))
```
Comment (lptr) [inline platforms/core-configuration/declarative-dsl-tooling-builders/build.gradle.kts:10]: 👏 That's what we want! Now let's get rid of that `:core` dependency too :D

File: platforms/core-runtime/service-provider/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+plugins {
+    id("gradlebuild.distribution.implementation-java")
+}
+
+description = "An API for providing internal services for Gradle modules"
+
+dependencies {
+    api(projects.javaLanguageExtensions)
```
Comment (lptr) [inline platforms/core-runtime/service-provider/build.gradle.kts:8]: ❓ How come we don't need `libs.inject`?

File: platforms/jvm/language-groovy/build.gradle.kts [hunk 3]
```diff
@@ -13,7 +13,6 @@ errorprone {
 }
 
 dependencies {
-    api(project(":base-services"))
```
Comment (lptr) [inline platforms/jvm/language-groovy/build.gradle.kts:16]: 👍 Yes!

**Review by lptr — APPROVED** [review]
LGTM!

File: platforms/core-runtime/service-provider/build.gradle.kts (hunk 2 above)
Comment (alllex) [reply platforms/core-runtime/service-provider/build.gradle.kts:10]: It's not required by the classes currently moved into this new module

**Comment by alllex** [comment]
@bot-gradle test without pts

**Comment by alllex** [comment]
@bot-gradle test without pts

File: platforms/core-runtime/service-provider/build.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,11 @@
+plugins {
+    id("gradlebuild.distribution.implementation-java")
+}
+
+description = "An API for providing internal services for Gradle modules"
```
Comment (alllex) [inline platforms/core-runtime/service-provider/build.gradle.kts:5]: ℹ️  Adding this new project is the main change

**Pull request merged**